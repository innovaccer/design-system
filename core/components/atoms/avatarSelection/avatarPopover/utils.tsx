import * as React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>,
  listRef?: any,
  withSearch?: boolean,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  triggerRef?: any
) => {
  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      navigateOptions('up', focusedOption, setFocusedOption, listRef, withSearch);
      break;
    case 'ArrowDown':
      event.preventDefault();
      navigateOptions('down', focusedOption, setFocusedOption, listRef, withSearch);
      break;
    case 'Enter':
      handleEnterKey(focusedOption);
      setHighlightLastItem?.(false);
      setHighlightFirstItem?.(false);
      break;
    case 'Escape':
      setOpenPopover?.(false);
      triggerRef.current.focus();
      setFocusedOption?.(undefined);
      break;
    default:
      break;
  }
};

const handleEnterKey = (focusedOption: Element | undefined) => {
  (focusedOption as HTMLElement)?.click();
};

const navigateOptions = (
  direction: string,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  listRef?: any,
  withSearch?: boolean
) => {
  const listItems = listRef?.current.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let index = Array.from(listItems).findIndex((item) => {
    return item == focusedOption;
  });

  if (index === -1) {
    index = direction === 'up' ? listItems.length - 1 : 0;
  } else if (
    (withSearch && index === 0 && direction === 'up') ||
    (withSearch && index === listItems.length - 1 && direction === 'down')
  ) {
    const searchInput = listRef.current.querySelector('[data-test="DesignSystem-AvatarSelection--Input"]');
    searchInput.focus();
    setFocusedOption && setFocusedOption(searchInput);
  } else {
    index = direction === 'up' ? (index - 1 + listItems.length) % listItems.length : (index + 1) % listItems.length;

    const targetOption = listItems[index];

    (targetOption as HTMLElement).focus();
    setFocusedOption && setFocusedOption(targetOption);
    targetOption.scrollIntoView({ block: 'center' });
  }
};

export const handleInputKeyDown = (
  event: React.KeyboardEvent,
  listRef: any,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  triggerRef?: any
) => {
  const listItems = listRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let targetOption;

  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      targetOption = listItems[listItems.length - 1];
      break;
    case 'ArrowDown':
      event.preventDefault();
      targetOption = listItems[0];
      break;
    case 'Escape':
      setOpenPopover?.(false);
      triggerRef.current.focus();
      setFocusedOption?.(undefined);
      break;
    default:
      break;
  }

  (targetOption as HTMLElement)?.focus();
  targetOption?.scrollIntoView?.({ block: 'center' });
  setFocusedOption && setFocusedOption(targetOption);
};
