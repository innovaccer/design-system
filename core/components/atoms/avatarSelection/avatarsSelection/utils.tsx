import * as React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>
) => {
  switch (event.key) {
    case 'Enter':
      setOpenPopover?.(true);
      setHighlightFirstItem?.(true);
      break;
    case 'ArrowDown':
      setOpenPopover?.(true);
      setHighlightFirstItem?.(true);
      break;
    case 'ArrowUp':
      setOpenPopover?.(true);
      setHighlightLastItem?.(true);
      break;
    default:
      break;
  }
};

export const focusListItem = (
  position: string,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  listRef?: any,
  withSearch?: boolean
) => {
  if (withSearch && position === 'down') {
    const searchInput = listRef.current?.querySelectorAll('[data-test="DesignSystem-AvatarSelection--Input"]');
    searchInput && searchInput[0]?.focus();
    setFocusedOption?.(searchInput);
    return;
  }

  const listItems = listRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let targetOption;

  if (position === 'down') {
    targetOption = listItems?.[0];
  } else {
    targetOption = listItems?.[listItems.length - 1];
  }
  (targetOption as HTMLElement)?.focus();
  targetOption?.scrollIntoView?.({ block: 'center' });
  setFocusedOption && setFocusedOption(targetOption);
};
