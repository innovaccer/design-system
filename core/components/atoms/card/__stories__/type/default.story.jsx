import { Card } from '@/index';

// CSF format story
export const defaultCard = () => {
  return (
    <Card>
      <div className="p-8" />
    </Card>
  );
};

export default {
  title: 'Components/Card/Type/Default Card',
  component: Card,
};
