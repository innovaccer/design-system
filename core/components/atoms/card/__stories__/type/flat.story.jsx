import { Card } from '@/index';

// CSF format story
export const Flat = () => {
  return (
    <Card shadow="none">
      <div className="p-8" />
    </Card>
  );
};

export default {
  title: 'Components/Card/Type/Flat',
  component: Card,
};
