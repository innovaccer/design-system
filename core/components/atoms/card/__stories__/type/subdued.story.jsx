import { Card, CardSubdued } from '@/index';

// CSF format story
export const Subdued = () => {
  return (
    <CardSubdued>
      <div className="p-8" />
    </CardSubdued>
  );
};

export default {
  title: 'Components/Card/Type/Subdued',
  component: Card,
  subcomponents: { CardSubdued },
  parameters: {
    docs: {
      docPage: {
        title: 'CardSubdued',
        description: 'A component to highlight section of a card.',
      },
    },
  },
};
