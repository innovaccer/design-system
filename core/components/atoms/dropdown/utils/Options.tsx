export let dropdownOptions: any[] = [];

for (let i = 1; i <= 50; i++) {
  dropdownOptions.push({
    label: `Option${i}`,
    value: `Option${i}`,
  });
}

export const selectedStoryOptions: any[] = [];

for (let i = 1; i <= 10; i++) {
  selectedStoryOptions.push({
    label: `Option${i}`,
    value: `Option${i}`,
    selected: i === 2,
  });
}

export const multiSelectedStoryOptions: any[] = [];

for (let i = 1; i <= 10; i++) {
  multiSelectedStoryOptions.push({
    label: `Option${i}`,
    value: `Option${i}`,
    selected: i === 2 || i === 7,
  });
}

export const storyOptions = dropdownOptions.slice(0, 10);

export const storySections = { 0: 'subheading', 5: 'subheading' };

export const subInfoItems = [
  {
    label: 'Option1',
    value: 'Option1',
    subInfo: 'subInfo'
  },
  {
    label: 'Option2',
    value: 'Option2',
    subInfo: 'subInfo'
  },
  {
    label: 'Option3',
    value: 'Option3',
    subInfo: 'subInfo',
  },
];

export const iconItems = [
  {
    label: 'Option1',
    value: 'Option1',
    icon: 'events'
  },
  {
    label: 'Option2',
    value: 'Option2',
    icon: 'events',
  },
  {
    label: 'Option3',
    value: 'Option3',
    icon: 'events',
  },
];
