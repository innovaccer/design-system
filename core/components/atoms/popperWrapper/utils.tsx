// Helper function to convert CSS class names to style objects
export const createStyleFromClass = (className: string) => {
  if (!className) return {};

  // Create a temporary element to get computed styles
  const tempElement = document.createElement('div');
  tempElement.className = className;
  document.body.appendChild(tempElement);

  const computedStyle = window.getComputedStyle(tempElement);
  const style: React.CSSProperties = {};

  // Extract relevant animation properties
  const animation = computedStyle.animation;
  const animationDuration = computedStyle.animationDuration;
  const animationTimingFunction = computedStyle.animationTimingFunction;
  const animationFillMode = computedStyle.animationFillMode;
  const opacity = computedStyle.opacity;
  const transform = computedStyle.transform;

  if (animation && animation !== 'none') {
    style.animation = animation;
  }
  if (animationDuration && animationDuration !== '0s') {
    style.animationDuration = animationDuration;
  }
  if (animationTimingFunction && animationTimingFunction !== 'ease') {
    style.animationTimingFunction = animationTimingFunction;
  }
  if (animationFillMode && animationFillMode !== 'none') {
    style.animationFillMode = animationFillMode;
  }
  if (opacity && opacity !== '1') {
    style.opacity = opacity;
  }
  if (transform && transform !== 'none') {
    style.transform = transform;
  }

  document.body.removeChild(tempElement);
  return style;
};
