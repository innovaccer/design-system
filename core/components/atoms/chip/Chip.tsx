import * as React from 'react';
import GenericChip from '../_chip';
import classNames from 'classnames';
import { BaseProps, extractBaseProps } from '@/utils/types';

export type ChipType = 'action' | 'selection' | 'input';
export type Name = number | string;

export interface ChipProps extends BaseProps {
  /**
   * Label of chip
   */
  label: string;
  /**
   * Type of material `Icon`
   */
  icon?: string;
  /**
   * Shows the 'clear' icon if value is not empty
   */
  clearButton?: boolean;
  /**
   * Disables the Chip, making it unable to be pressed
   */
  disabled?: boolean;
  /**
   * Select the chip
   */
  selected?: boolean;
  /**
   * Type of chip
   */
  type: ChipType;
  /**
   * Handler to be called when Chip is closed
   */
  onClose?: (name: Name) => void;
  /**
   * Handler to be called when Chip is clicked
   */
  onClick?: (name: Name) => void;
  name: Name;
}

export const Chip = (props: ChipProps) => {
  const { label, icon, clearButton, type, disabled, selected, onClose, onClick, name, className } = props;

  const baseProps = extractBaseProps(props);

  const onCloseHandler = () => {
    if (!disabled && onClose) onClose(name);
  };
  const onClickHandler = () => {
    if (!disabled && onClick) onClick(name);
  };

  const chipClass = classNames(
    {
      Chip: true,
      [`Chip-${type}--disabled`]: disabled,
      [`Chip--${type}`]: type && !disabled,
      [`Chip-${type}--selected`]: selected && !disabled,
      [`Chip-selection--selectedDisabled`]: type === 'selection' && selected && disabled,
    },
    className
  );

  const clearbutton = type === 'action' ? false : clearButton;
  const select = type === 'selection' && selected ? true : false;

  return (
    <GenericChip
      data-test="DesignSystem-Chip--GenericChip"
      {...baseProps}
      label={label}
      selected={select}
      icon={icon}
      clearButton={clearbutton}
      disabled={disabled}
      className={chipClass}
      onClose={onCloseHandler}
      onClick={onClickHandler}
      name={name}
    />
  );
};

Chip.displayName = 'Chip';
Chip.defaultProps = {
  type: 'input',
};

export default Chip;
