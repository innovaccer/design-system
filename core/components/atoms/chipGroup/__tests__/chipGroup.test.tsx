import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { testHelper, filterUndefined, valueHelper, testMessageHelper } from '@/utils/testHelper';
import { ChipGroup } from '@/index';
import { ChipGroupProps as Props } from '@/index.type';
import { Type } from '../../chip/Chip';

const type: Type[] = ['action', 'input', 'selection'];
const FunctionValue = jest.fn();
const onClick = jest.fn();
const onClose = jest.fn();

const list =
  [
    {
      label: 'Action', icon: 'assessment', clearButton: true, disabled: false,
      name: '1', type: type[0],
    },
    {
      label: 'Input', icon: 'assessment', clearButton: true, disabled: false,
      name: '2', type: type[1],
    },
    {
      label: 'Selection', icon: 'assessment', clearButton: true, disabled: false,
      selected: true, name: '3', type: type[2],
    },
    {
      label: 'Selection', icon: 'assessment', clearButton: true, disabled: false,
      name: '4', type: type[2],
    }
  ];

describe('ChipGroup component', () => {
  const mapper: Record<string, any> = {
    list: valueHelper(list, { required: true }),
    onclick: valueHelper(onClick, { required: true }),
    onClose: valueHelper(onClose, { required: true }),

  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as Props;

    it(testMessageHelper(attr), () => {
      const { baseElement } = render(
            <ChipGroup
              {...attr}
            />
        );
      expect(baseElement).toMatchSnapshot();
    });
  };

  testHelper(mapper, testFunc);
});

describe('ChipGroup component', () => {

  it('renders chipGroup component', () => {
    const { getByTestId, getAllByTestId } = render(
      <ChipGroup
        list={list}
      />
    );
    expect(getByTestId('DesignSystem-ChipGroup')).toBeInTheDocument();
    expect(getAllByTestId('DesignSystem-ChipGroup--Chip')).toHaveLength(list.length);
  });

  it('renders ChipGroup component with prop onClick', () => {
    const { getAllByTestId } = render(
      <ChipGroup
        list={list}
        onClick={FunctionValue}
      />
    );
    const onClickFunction = getAllByTestId('DesignSystem-ChipGroup--Chip')[0];
    fireEvent.click(onClickFunction);
    expect(FunctionValue).toHaveBeenCalled();
  });

  it('renders ChipGroup component with prop onClose', () => {
    const { getAllByTestId } = render(
      <ChipGroup
        list={list}
        onClose={FunctionValue}
      />
    );
    const onCloseFunction = getAllByTestId('DesignSystem-GenericChip--clearButton')[0];
    fireEvent.click(onCloseFunction);
    expect(FunctionValue).toHaveBeenCalled();
  });

});
