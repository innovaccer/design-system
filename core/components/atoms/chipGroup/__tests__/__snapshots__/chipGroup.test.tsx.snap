// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChipGroup component 
  list: [{"label":"Action","icon":"assessment","clearButton":true,"disabled":false,"name":"1","type":"action"},{"label":"Input","icon":"assessment","clearButton":true,"disabled":false,"name":"2","type":"input"},{"label":"Selection","icon":"assessment","clearButton":true,"disabled":false,"selected":true,"name":"3","type":"selection"},{"label":"Selection","icon":"assessment","clearButton":true,"disabled":false,"name":"4","type":"selection"}], onclick: "[Function]", onClose: "[Function]"
 1`] = `
<body>
  <div>
    <div
      class="ChipGroup"
      data-test="DesignSystem-ChipGroup"
    >
      <span
        class="ChipGroup-item"
      >
        <div
          class="Chip-wrapper Chip Chip--action"
          data-test="DesignSystem-ChipGroup--Chip"
        >
          <i
            class="material-icons material-icons-round Icon Icon--default Chip-icon Chip-icon--left"
            data-test="DesignSystem-GenericChip--Icon"
            role="button"
            style="font-size: 16px; width: 16px;"
          >
            assessment_round
          </i>
          <span
            class="Text Text--default Text--regular"
            data-test="DesignSystem-GenericChip--Text"
          >
            Action
          </span>
        </div>
      </span>
      <span
        class="ChipGroup-item"
      >
        <div
          class="Chip-wrapper Chip Chip--input"
          data-test="DesignSystem-ChipGroup--Chip"
        >
          <i
            class="material-icons material-icons-round Icon Chip-icon Chip-icon--left"
            data-test="DesignSystem-GenericChip--Icon"
            role="button"
            style="font-size: 16px; width: 16px;"
          >
            assessment_round
          </i>
          <span
            class="Text Text--default Text--regular"
            data-test="DesignSystem-GenericChip--Text"
          >
            Input
          </span>
          <i
            class="material-icons material-icons-round Icon Chip-icon Chip-icon--right cursor-pointer"
            data-test="DesignSystem-GenericChip--clearButton"
            role="button"
            style="font-size: 16px; width: 16px;"
            tabindex="0"
          >
            clear_round
          </i>
        </div>
      </span>
      <span
        class="ChipGroup-item"
      >
        <div
          class="Chip-wrapper Chip Chip--selection Chip-selection--selected"
          data-test="DesignSystem-ChipGroup--Chip"
        >
          <i
            class="material-icons material-icons-round Icon Icon--info Chip-icon Chip-icon--left"
            data-test="DesignSystem-GenericChip--Icon"
            role="button"
            style="font-size: 16px; width: 16px;"
          >
            assessment_round
          </i>
          <span
            class="Text Text--link Text--regular"
            data-test="DesignSystem-GenericChip--Text"
          >
            Selection
          </span>
          <i
            class="material-icons material-icons-round Icon Icon--info Chip-icon Chip-icon--right cursor-pointer"
            data-test="DesignSystem-GenericChip--clearButton"
            role="button"
            style="font-size: 16px; width: 16px;"
            tabindex="0"
          >
            clear_round
          </i>
        </div>
      </span>
      <span
        class="ChipGroup-item"
      >
        <div
          class="Chip-wrapper Chip Chip--selection"
          data-test="DesignSystem-ChipGroup--Chip"
        >
          <i
            class="material-icons material-icons-round Icon Chip-icon Chip-icon--left"
            data-test="DesignSystem-GenericChip--Icon"
            role="button"
            style="font-size: 16px; width: 16px;"
          >
            assessment_round
          </i>
          <span
            class="Text Text--default Text--regular"
            data-test="DesignSystem-GenericChip--Text"
          >
            Selection
          </span>
          <i
            class="material-icons material-icons-round Icon Chip-icon Chip-icon--right cursor-pointer"
            data-test="DesignSystem-GenericChip--clearButton"
            role="button"
            style="font-size: 16px; width: 16px;"
            tabindex="0"
          >
            clear_round
          </i>
        </div>
      </span>
    </div>
  </div>
</body>
`;
